keps/sig-network/1880-services-ip-ranges/README.md

```diff
@@ -0,0 +1,935 @@
+
+ - a ServiceIPRange will be immutable after creation
+ - a ServiceIPRange can only be deleted if there are no ServiceIP associated to it
+ - there can not be overlapping IP networks on the ServiceIPRanges created
+ - there has to be always an unique and default ServiceIPRange
+ - any ServiceIP existing in a cluster has to belong to a ServiceIPRange
+ - any Service with a ClusterIP assigned has to have always a ServiceIP object associated
+
+### User Stories
+
+#### Story 1
+
+As a Kubernetes user I want to be able to dynamically increase the number of IPs availables for
+Services.
+
+#### Story 2
+
+As a Kubernetes admin I want to have a process that allows me to renumber my Services IPs.
+
+#### Story 3
+
+As a Kubernetes developer I want to be able to know the current Service IP range and the IP
+addresses allocated.
+
+#### Story 4
+
+As a Kubernetes admin that wants to use IPv6, I want to be able to follow the IPv6 address planning
+best practices, being able to assign /64 prefixes to my end subnets.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+Current API machinery doesn't consider transactions, Services API is kind of special in this regard
+since already performs allocations inside the Registry pipeline using the Storage to keep
+consistency on an Obscure API Object that stores the bitmap with the allocations.
+
+This proposal also requires to perform validation from one Object against other Object inside an API
+request.
+
+### Risks and Mitigations
+
+Cross validation of Objects is not common in the Kubernetes API, sig-apimachinery should verify not
+bad precedent is introduced.
+
+Current Kubernetes cluster have a very static network configuration, allowing to expand the Service
+IP ranges will give more flexibility to users, with the risk of having problematic or incongruent
+network configurations with overlapping. But this is not really a new problem, users need to do a
+proper network planning before adding new ranges to the Service IP pool.
+
+## Design Details
+
+### Current implementation details 
+
+[Kubernetes
+Services](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#service-v1-core) need
+to be able to dynamically allocate resources from a predefined range/pool for ClusterIPs.
+
+ClusterIP is the IP address of the service and is usually assigned randomly. If an address is
+specified manually, is in-range, and is not in use, it will be allocated to the service; otherwise
+creation of the service will fail. The Service ClusterIP range is defined in the kube-apiserver with
+the following flags:
+
+```
+--service-cluster-ip-range string
+A CIDR notation IP range from which to assign service cluster IPs. This must not overlap with any IP ranges assigned to nodes or pods. Max of two dual-stack CIDRs is allowed.
+```
+
+And in the controller-manager:
+
+```
+--service-cluster-ip-range string
+CIDR Range for Services in cluster. Requires --allocate-node-cidrs to be true
```

Comment:
the controller-manager only uses the service cluster ip range for the ipam controller, it just skip the service ranges from the node ranges to avoid overlapping between services and pods